src/v/kafka/server/snc_quota_manager.cc

```diff
@@ -318,7 +381,14 @@ void snc_quota_manager::record_request_receive(
     if (ctx._exempt) {
         return;
     }
-    _shard_quota.in.use(request_size, now);
+    if (_use_throttling_v2()) {
+        if (_node_quota.in) {
+            _node_quota.in->replenish(now);
+            _node_quota.in->grab(request_size);
```

Comment:
So given how the rovers in the shared_token_bucket work and given that we unconditionally grab from the bucket I think we can end up in a situation where:

 - A large burst causes tail >> head
 - Incoming rate then equals replenish rate / limit
 - All requests continue to get dropped as tail >= head is a continuous state 

Am I missing something? Is that the intended behaviour?

Reply:
I guess assuming everyone fully respects the throttle time it will work fine.

Reply:
Yeah, throttling is enforced (on the second read), so as long as the set of outstanding requests doesn't overflow a uint64_t, it should all work.